raster-foundry/raster-foundry
# Fetch tag refs during Jenkins SCM checkout

## Overview

Builds have been failing since we merged #4222 (which uncomments publishing snapshot artifacts). 

This is because we use the output of  `$ git describe` as the version for published snapshot artifacts, and that depends on tags being present.

>Multi-branch pipeline jobs (as of git plugin 3.4.0) intentionally use a narrow refspec targeted specifically at the branch being built, and it intentionally does not fetch tags. The change was made because the typical use case for the pipeline job does not benefit from the extra effort to use a broad refspec and to fetch tags.

I believe the `java.util.NoSuchElementException: None.get` error can be traced to this line:

```groovy
addCommandAlias(
  "gitSnapshots",
  ";set version in ThisBuild := git.gitDescribedVersion.value.get + \"-SNAPSHOT\"")
```

Some builds are passing because I ran the equivalent of `$ git fetch -t` in the workspace of the first build executor. It looks like the default `checkout scm` task doesn't clear these tags. Basically, builds picked up by the first executor are passing, but builds picked up by the second executor are failing:

```bash
ubuntu@ip-172-31-52-80:/var/lib/jenkins/workspace$ cd ndry_raster-foundry_develop-5ZSVDGSBLEMMDMUZZFN5PIIJBEKWX3XWPZE7V6RMSGTJZZULUF6A
ubuntu@ip-172-31-52-80:/var/lib/jenkins/workspace/ndry_raster-foundry_develop-5ZSVDGSBLEMMDMUZZFN5PIIJBEKWX3XWPZE7V6RMSGTJZZULUF6A$ git describe
1.13.0-15-ge00bf87
```
```bash
ubuntu@ip-172-31-52-80:/var/lib/jenkins/workspace$ cd ndry_raster-foundry_develop-5ZSVDGSBLEMMDMUZZFN5PIIJBEKWX3XWPZE7V6RMSGTJZZULUF6A@2
ubuntu@ip-172-31-52-80:/var/lib/jenkins/workspace/ndry_raster-foundry_develop-5ZSVDGSBLEMMDMUZZFN5PIIJBEKWX3XWPZE7V6RMSGTJZZULUF6A@2$ git describe
fatal: No names found, cannot describe anything.
```

This PR configures the `checkout scm` task to fetch tags.

See: https://github.com/allegro/axion-release-plugin/issues/195#issuecomment-319541175

## Notes

This requires [script approvals](http://jenkins.staging.rasterfoundry.com/scriptApproval/). The only way I was able to add them was to run the build several times until I could approve each permission.

It's very strange this is happening, because I remember working through this exact issue with @tnation14. I don't know why our changes didn't make it into #3873.

## Testing Instructions

Stage the same workspace as [build 889](http://jenkins.staging.rasterfoundry.com/job/Raster%20Foundry/job/raster-foundry/job/develop/899/consoleFull):

```bash
$ git init
$ git config remote.origin.url https://github.com/raster-foundry/raster-foundry.git
$ git fetch --no-tags --progress https://github.com/raster-foundry/raster-foundry.git +refs/heads/develop:refs/remotes/origin/develop
$ git config core.sparsecheckout
$ git checkout -f e129b95c4218bd6c8716b8bbce44e94a311e7b26
```

Ensure build failure:

```bash
$ AWS_PROFILE=raster-foundry aws s3 cp "s3://rasterfoundry-testing-config-us-east-1/.env" ".env"
$ docker-compose \
    -f docker-compose.test.yml \
    run --rm --no-deps \
    -e SONATYPE_USERNAME=placeholder \
    -e SONATYPE_PASSWORD=placeholder \
    -e PGP_HEX_KEY="555555" \
    -e PGP_PASSPHRASE=placeholder \
    build gitSnapshots api/publishSigned
```

Ensure build succeeds past `Applying settings...`:

```bash
$ git fetch -t
```

```bash
$ AWS_PROFILE=raster-foundry aws s3 cp "s3://rasterfoundry-testing-config-us-east-1/.env" ".env"
$ docker-compose \
    -f docker-compose.test.yml \
    run --rm --no-deps \
    -e SONATYPE_USERNAME=placeholder \
    -e SONATYPE_PASSWORD=placeholder \
    -e PGP_HEX_KEY="555555" \
    -e PGP_PASSPHRASE=placeholder \
    build gitSnapshots api/publishSigned
```
